packages/node/src/lib/get-project-base-url.ts

```diff
@@ -0,0 +1,75 @@
+import crypto from 'crypto';
+
+import findCacheDir from 'find-cache-dir';
+import flatCache from 'flat-cache';
+import fetch from 'node-fetch';
+import timeoutSignal from 'timeout-signal';
+
+import pkg from '../../package.json';
+import config from '../config';
+
+export function getCache(readmeApiKey) {
+  const encodedApiKey = Buffer.from(`${readmeApiKey}:`).toString('base64');
+  const cacheDir = findCacheDir({ name: pkg.name, create: true });
+  const fsSafeApikey = crypto.createHash('md5').update(encodedApiKey).digest('hex');
+
+  // Since we might have differences of cache management, set the package version into the cache key so all caches will
+  // automatically get refreshed when the package is updated/installed.
+  const cacheKey = `${pkg.name}-${pkg.version}-${fsSafeApikey}`;
+
+  return flatCache.load(cacheKey, cacheDir);
+}
+
+export async function getProjectBaseUrl(readmeApiKey: string, requestTimeout = config.timeout): Promise<string> {
+  const encodedApiKey = Buffer.from(`${readmeApiKey}:`).toString('base64');
+
+  const cache = getCache(readmeApiKey);
+  // Does the cache exist? If it doesn't, let's fill it. If it does, let's see if it's stale. Caches should have a TTL
+  // of 1 day.
+  const lastUpdated = cache.getKey('lastUpdated');
+
+  if (
+    lastUpdated === undefined ||
+    (lastUpdated !== undefined && Math.abs(lastUpdated - Math.round(Date.now() / 1000)) >= 86400)
+  ) {
+    const signal = timeoutSignal(requestTimeout);
+
+    let baseUrl;
+    await fetch(`${config.readmeApiUrl}/v1/`, {
+      method: 'get',
+      headers: {
+        Authorization: `Basic ${encodedApiKey}`,
+        'User-Agent': `${pkg.name}/${pkg.version}`,
+      },
+      signal,
+    })
+      .then(res => {
+        if (res.status >= 400 && res.status <= 599) {
+          throw res;
+        }
+
+        return res.json();
+      })
+      .then(project => {
+        baseUrl = project.baseUrl;
+
+        cache.setKey('baseUrl', project.baseUrl);
+        cache.setKey('lastUpdated', Math.round(Date.now() / 1000));
+      })
+      .catch(() => {
+        // If unable to access the ReadMe API for whatever reason, let's set the last updated time to two minutes from
+        // now yesterday so that in 2 minutes we'll automatically make another attempt.
+        cache.setKey('baseUrl', null);
+        cache.setKey('lastUpdated', Math.round(Date.now() / 1000) - 86400 + 120);
+      })
+      .finally(() => {
+        timeoutSignal.clear(signal);
+      });
+
+    cache.save();
+
+    return baseUrl;
+  }
+
+  return cache.getKey('baseUrl');
```

Comment:
They actually don't any more, since I added a conditional around whether we need to create the promise or not: https://github.com/readmeio/metrics-sdks/blob/151bb1db30fc5ad646831d50c795e5f9dc460643/packages/node/src/lib/log.ts#L108-L116 

So getProjectBaseUrl is itself an async function that has an await in it, then after that await, it just returns a value which in turn resolves the upper promise! I can't quite remember the circumstances in that other PR why I had to add Promise.resolve to them all.

Reply:
makes sense!